features/steps/component.go

```diff
@@ -2,235 +2,100 @@ package steps
+// as a result of test setup, are picked up
+func (c *Component) Router() (http.Handler, error) {
+	if err := c.svc.Init(context.Background(), c.svc.Cfg, BuildTime, GitCommit, Version); err != nil {
+		c.ErrorFeature.Fatalf("failed to initialise service: %s", err)
 	}
 
-	c.cfg.Mongo.Database = utils.RandomDatabase()
+	return c.svc.Api.Router, nil
 }
 
-// startService starts the service under test and blocks until an error or an os interrupt is received.
-// Then it closes the service (graceful shutdown)
-func (c *Component) startService(ctx context.Context) {
-	defer c.wg.Done()
-	c.svc.Start(ctx, c.errorChan)
-
-	wg := sync.WaitGroup{}
-	if err := c.drainTopic(c.ctx, c.cfg.KafkaConfig.ExportStartTopic, ComponentTestGroup, &wg); err != nil {
-		log.Error(c.ctx, "error draining topic", err)
-	}
-
-	select {
-	case err := <-c.errorChan:
-		err = fmt.Errorf("service error received: %w", err)
-		defer func() {
-			if err := c.svc.Close(ctx); err != nil {
-				log.Error(ctx, "failed to shutdown service gracefully: %s", err)
-			}
-		}()
-		panic(fmt.Errorf("unexpected error received from errorChan: %w", err))
-	case sig := <-c.signals:
-		log.Info(ctx, "os signal received", log.Data{"signal": sig})
-	}
-	if err := c.svc.Close(ctx); err != nil {
-		panic(fmt.Errorf("failed to shutdiwn gracefully: %w", err))
-	}
+// Reset re-initialises the service under test and the api mocks.
+func (c *Component) Reset() {
+	c.AuthFeature.Reset()
+	c.APIFeature.Reset()
+	c.DatasetFeature.Reset()
+	c.CantabularFeature.Reset()
+	c.MongoFeature.Reset()
 }
 
-// Close kills the application under test, and then it shuts down the testing producer.
 func (c *Component) Close() {
-
-	if !c.shutdownInitiated {
-		c.shutdownInitiated = true
-		c.signals <- os.Interrupt
-
-		// wait for graceful shutdown to finish (or timeout)
-		// TODO we should fix the timeout issue and then uncomment the following line.
-		c.wg.Wait()
-	}
-
+	c.AuthFeature.Close()
+	c.DatasetFeature.Close()
+	c.MongoFeature.Close()
 }
 
-// Reset re-initialises the service under test and the api mocks.
-// Note that the service under test should not be started yet
-// to prevent race conditions if it tries to call un-initialised dependencies (steps)
-func (c *Component) Reset() error {
-	c.setInitialiserMock()
-	c.DatasetAPI.Reset()
-	c.CantabularClient.Reset()
-
-	if _, err := c.Init(); err != nil {
-		return fmt.Errorf("failed to initialise component: %w", err)
-	}
-
-	return nil
-}
-
-func GetWorkingMongo(ctx context.Context, cfg *config.Config, g service.Generator) (service.Datastore, error) {
-	mongoClient, err := mongodb.NewClient(ctx, g, mongodb.Config{
-		MongoDriverConfig:       cfg.Mongo,
-		FilterFlexAPIURL:        cfg.BindAddr,
-		FiltersCollection:       cfg.FiltersCollection,
-		FilterOutputsCollection: cfg.FilterOutputsCollection,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to create new mongo mongoClient: %w", err)
+func setInitialiserMock(g service.Generator) {
```

Comment:
Why remove this function from the component struct? Just means you have to pass extra arguments now

Reply:
I have added this function back as a method on component, which combines what we talked about yesterday re: features setting their service injection points. The reason 'g' is passed as a parameter is to avoid cluttering the component object with unnecessary state - after initialisation g is never accessed by the component object again.